File: plugin/src/main/java/com/jtprince/bingo/plugin/Messages.java
package com.jtprince.bingo.plugin;

import com.jtprince.bingo.plugin.player.BingoPlayer;
import com.jtprince.bingo.plugin.player.BingoPlayerRemote;
import com.jtprince.bingo.plugin.player.BingoPlayerTeam;
import com.jtprince.util.ChatUtils;
import net.kyori.adventure.audience.Audience;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import net.kyori.adventure.text.event.ClickEvent;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextColor;
import net.kyori.adventure.text.format.TextDecoration;
import org.bukkit.entity.Player;

import java.net.URL;
import java.util.Collection;
import java.util.HashSet;
import java.util.stream.Collectors;

public class Messages {
    final BingoGame game;

    private static final TextColor COLOR_HEADER = NamedTextColor.GOLD;
    private static final TextColor COLOR_TEXT = NamedTextColor.AQUA;

    private static final TextComponent HEADER_NO_GAME = Component.text("[BINGO]")
            .color(COLOR_HEADER).decoration(TextDecoration.BOLD, true);

    public Messages(BingoGame game) {
        this.game = game;
    }

    private TextComponent withHeader(TextComponent component) {
        if (this.game.gameCode != null) {
            return Component.empty().color(COLOR_TEXT)
                .append(HEADER_NO_GAME.hoverEvent(Component.text("Game Code: " + game.gameCode)))
                .append(Component.space())
                .append(component);
        } else {
            return withNoGameHeader(component);
        }
    }

    private static TextComponent withNoGameHeader(TextComponent component) {
        return Component.empty().color(COLOR_TEXT)
            .append(HEADER_NO_GAME)
            .append(Component.space())
            .append(component);
    }

    private void sendWithHeader(Audience a, TextComponent component) {
        a.sendMessage(withHeader(component));
    }

    private void announceWithHeader(TextComponent component) {
        MCBingoPlugin.instance().getServer().sendMessage(withHeader(component));
    }

    private static void sendGamelessWithHeader(Audience a, TextComponent component) {
        a.sendMessage(withNoGameHeader(component));
    }

    /* Begin external facing calls */

    public void basicTell(Audience a, String msg) {
        sendWithHeader(a, Component.text(msg));
    }

    public void basicAnnounce(String msg) {
        sendWithHeader(MCBingoPlugin.instance().getServer(), Component.text(msg));
    }

    public static void basicTellNoGame(Audience a, String msg) {
        sendGamelessWithHeader(a, Component.text(msg));
    }

    public void announcePreparingGame() {
        TextComponent component = Component
            .text("Generating worlds for new game ")
            .append(Component.text(game.gameCode).color(NamedTextColor.BLUE))
            .append(Component.text("."));
        announceWithHeader(component);

        component = Component.text("This will cause the server to lag!");
        announceWithHeader(component);
    }

    public void announceGameFailed() {
        TextComponent component = Component
            .text("Failed to connect to the Bingo board.")
            .color(NamedTextColor.RED);
        announceWithHeader(component);
    }

    public void announceWorldsGenerated(Collection<BingoPlayer> players) {
        players = new HashSet<>(players);
        players.add(new BingoPlayerRemote("Hello world"));
        TextComponent playersCpnt = ChatUtils.commaSeparated(
            players.stream().map(BingoPlayer::getFormattedName)
                .collect(Collectors.toUnmodifiableSet())
        );
        TextComponent component = Component
            .text("Bingo worlds have been generated for ")
            .append(Component.text(players.size() + " players.").hoverEvent(playersCpnt));

        announceWithHeader(component);
    }

    public void announceGameReady(Collection<BingoPlayer> players) {
        TextComponent component;
        for (BingoPlayer p : players) {
            // Game link for this specific player
            URL url = MCBConfig.getGameUrl(this.game.gameCode, p);
            if (url == null) {
                component = Component
                    .text("Could not get the board to link you to! Contact the server admin "
                        + "to update the plugin's config.yml.")
                    .color(NamedTextColor.RED);
            } else {
                component = Component.empty()
                    .append(Component.text("[Open Board]")
                        .decoration(TextDecoration.UNDERLINED, true)
                        .color(NamedTextColor.YELLOW)
                        .hoverEvent(Component.text(url.toString()))
                        .clickEvent(ClickEvent.openUrl(url))
                    )
                    .append(Component.space())
                    .append(Component.text("[START]")
                        .decoration(TextDecoration.UNDERLINED, true)
                        .color(NamedTextColor.GREEN)
                        .clickEvent(ClickEvent.runCommand("/bingo start"))
                        .hoverEvent(Component.text("Start Game (/bingo start)"))
                    );
            }

            for (Player bukkitPlayer : p.getBukkitPlayers()) {
                sendWithHeader(bukkitPlayer, component);
            }
        }
    }

    public void announcePlayerMarking(BingoPlayer player, Space space, boolean invalidated) {
        TextComponent component;
        if (!invalidated) {
            component = Component.empty()
                .append(player.getFormattedName())
                .append(Component.text(" has marked "))
                .append(Component.text(space.text).color(NamedTextColor.GREEN))
                .append(Component.text("!"));
        } else {
            component = Component.empty()
                .append(player.getFormattedName())
                .append(Component.text(" has invalidated "))
                .append(Component.text(space.text).color(NamedTextColor.RED))
                .append(Component.text("!"));
        }

        announceWithHeader(component);
    }

    public void tellPlayerTeams(Collection<BingoPlayer> players) {
        for (BingoPlayer bp : players) {
            if (bp instanceof BingoPlayerTeam) {
                BingoPlayerTeam bpt = (BingoPlayerTeam) bp;

                Component component = Component.text("You are playing on team ")
                    .append(bpt.getFormattedName())
                    .append(Component.text("!"));
                Audience.audience(bpt.getBukkitPlayers()).sendMessage(component);
            }
        }
    }
}

File: kotlinplugin/src/main/kotlin/com/jtprince/bingo/kplugin/Messages.kt
package com.jtprince.bingo.kplugin

import com.jtprince.bingo.kplugin.game.BingoGame
import com.jtprince.bingo.kplugin.player.BingoPlayer
import com.jtprince.bingo.kplugin.player.BingoPlayerTeam
import com.jtprince.util.ChatUtils
import io.ktor.http.*
import net.kyori.adventure.audience.Audience
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.event.ClickEvent
import net.kyori.adventure.text.event.HoverEvent
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextColor
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.Bukkit
import java.net.URL

object Messages {
    private val COLOR_HEADER: TextColor = NamedTextColor.GOLD
    private val COLOR_TEXT: TextColor = NamedTextColor.AQUA

    private val HEADER = Component.text("[BINGO]")
        .color(COLOR_HEADER).decoration(TextDecoration.BOLD, true)

    private fun Component.withBingoHeader(): Component {
        val gameCode = BingoGame.currentGame?.gameCode
        var builder = Component.empty().color(COLOR_TEXT)
        builder = if (gameCode != null) {
            builder.append(HEADER.hoverEvent(HoverEvent.showText(Component.text(
                "Game Code: $gameCode"
            ))))
        } else {
            builder.append(HEADER)
        }
        return builder.append(Component.space()).append(this)
    }

    private fun sendWithHeader(a: Audience, component: Component) {
        a.sendMessage(component.withBingoHeader())
    }

    private fun announceWithHeader(component: Component) {
        Bukkit.getServer().sendMessage(component.withBingoHeader())
    }

    /* Begin external facing calls */

    fun Audience.bingoTell(msg: String) {
        bingoTell(Component.text(msg))
    }

    fun Audience.bingoTellError(msg: String) {
        bingoTell(Component.text(msg).color(NamedTextColor.RED))
    }

    fun Audience.bingoTell(msg: Component) {
        sendWithHeader(this, msg)
    }

    fun bingoAnnounce(msg: String) {
        bingoAnnounce(Component.text(msg))
    }

    fun bingoAnnounce(msg: Component) {
        announceWithHeader(msg)
    }

    fun bingoAnnouncePreparingGame(gameCode: String) {
        val component = Component
            .text("Generating worlds for new game ")
            .append(Component.text(gameCode).color(NamedTextColor.BLUE))
            .append(Component.text("."))
        announceWithHeader(component)
        Bukkit.getServer().sendActionBar(
            Component.text("Generating Worlds - Server will lag!")
                .color(NamedTextColor.GOLD)
        )
    }

    fun bingoAnnounceGameFailed() {
        val component = Component
            .text("Failed to connect to the Bingo board.")
            .color(NamedTextColor.RED)
        announceWithHeader(component)
    }

    fun bingoAnnounceWorldsGenerated(players: Collection<BingoPlayer>) {
        val playersCpnt = ChatUtils.commaSeparated(
            players.map(BingoPlayer::formattedName))
        val component = Component
            .text("Bingo worlds have been generated for ")
            .append(
                Component.text("${players.size} player${if (players.size != 1) 's' else ""}.")
                    .hoverEvent(playersCpnt)
            )
        announceWithHeader(component)
    }

    fun bingoAnnounceGameReady(gameCode: String, players: Collection<BingoPlayer>, starters: Audience) {
        for (p in players) {
            // Game link for this specific player
            val url: Url = BingoConfig.gameUrl(gameCode, p)
            val component = Component.empty()
                .append(
                    Component.text("[Open Board]")
                        .decoration(TextDecoration.UNDERLINED, true)
                        .color(NamedTextColor.YELLOW)
                        .hoverEvent(Component.text(url.toString()))
                        .clickEvent(ClickEvent.openUrl(URL(url.toString())))
                )
            for (bukkitPlayer in p.bukkitPlayers) {
                sendWithHeader(bukkitPlayer, component)
            }
        }
        val component = Component.text("[START]")
            .decoration(TextDecoration.UNDERLINED, true)
            .color(NamedTextColor.GREEN)
            .clickEvent(ClickEvent.runCommand("/bingo start"))
            .hoverEvent(Component.text("Start Game (/bingo start)"))
        sendWithHeader(starters, component)
    }

    fun Audience.bingoTellGoalProgress(progress: Int) {
        val component = Component.text("Progress: $progress")
            .color(NamedTextColor.GRAY)
            .decorate(TextDecoration.ITALIC)
        sendWithHeader(this, component)
    }

    fun bingoAnnouncePlayerMarking(player: BingoPlayer, spaceText: String, invalidated: Boolean) {
        val component = if (!invalidated) {
            Component.empty()
                .append(player.formattedName)
                .append(Component.text(" has marked "))
                .append(Component.text(spaceText).color(NamedTextColor.GREEN))
                .append(Component.text("!"))
        } else {
            Component.empty()
                .append(player.formattedName)
                .append(Component.text(" has invalidated "))
                .append(Component.text(spaceText).color(NamedTextColor.RED))
                .append(Component.text("!"))
        }
        announceWithHeader(component)
    }

    fun bingoAnnounceEnd(winner: BingoPlayer?) {
        bingoAnnounce(Component.text("The game has ended!"))
        winner?.also {
            val component = Component.empty()
                .color(NamedTextColor.GOLD)
                .append(it.formattedName)
                .append(Component.text(" is the winner!"))
            announceWithHeader(component)
        }
    }

    fun Audience.bingoTellNotReady() {
        bingoTellError("The game is not ready to be started!")
    }

    fun bingoTellTeams(players: Collection<BingoPlayer>) {
        for (bp in players) {
            if (bp is BingoPlayerTeam) {
                val component = Component.text("You are playing on team ")
                    .append(bp.formattedName)
                    .append(Component.text("!"))
                Audience.audience(bp.bukkitPlayers).sendMessage(component)
            }
        }
    }
}
